crypto/fipsmodule/service_indicator/service_indicator.c

```diff
@@ -0,0 +1,92 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <openssl/crypto.h>
+#include <openssl/service_indicator.h>
+
+#if defined(AWSLC_FIPS)
+int awslc_fips_service_indicator_get_counter(void) {
+  struct fips_service_indicator_state *indicator =
+      CRYPTO_get_thread_local(AWSLC_THREAD_LOCAL_FIPS_SERVICE_INDICATOR_STATE);
+  if (!indicator) {
+    awslc_fips_service_indicator_init_state();
+    indicator = CRYPTO_get_thread_local(AWSLC_THREAD_LOCAL_FIPS_SERVICE_INDICATOR_STATE);
+  }
+  return indicator->counter;
+}
+
+struct fips_service_indicator_state *awslc_fips_service_indicator_get_state(void) {
+  struct fips_service_indicator_state *indicator =
+      CRYPTO_get_thread_local(AWSLC_THREAD_LOCAL_FIPS_SERVICE_INDICATOR_STATE);
+  if(!indicator) {
+    awslc_fips_service_indicator_init_state();
+    indicator = CRYPTO_get_thread_local(AWSLC_THREAD_LOCAL_FIPS_SERVICE_INDICATOR_STATE);
+  }
+  struct fips_service_indicator_state *status = OPENSSL_malloc(sizeof(struct fips_service_indicator_state));
+  OPENSSL_memset(status, 0, sizeof(struct fips_service_indicator_state));
+  OPENSSL_memcpy(status, indicator, sizeof(struct fips_service_indicator_state));
+  return status;
+}
+
+void awslc_fips_service_indicator_reset_state(void) {
+  struct fips_service_indicator_state *indicator =
+      CRYPTO_get_thread_local(AWSLC_THREAD_LOCAL_FIPS_SERVICE_INDICATOR_STATE);
+  if (!indicator) {
+    awslc_fips_service_indicator_init_state();
+    indicator = CRYPTO_get_thread_local(AWSLC_THREAD_LOCAL_FIPS_SERVICE_INDICATOR_STATE);
+  }
+  indicator->counter = 0;
+  indicator->serviceID = fips_approved_no_state;
+}
+
+int awslc_fips_check_service_approved(int prev_counter, int service_id) {
```

Comment:
I was thinking we could have a variadic macro which calls the function and checks if it's approved, for example:
```
#define IS_FIPS_APPROVED_CALL_SERVICE(approved, func, ...)  \
  do { \
    approved = 0 \
    counter = awslc_fips_service_indicator_get_counter() \
    func(__VA_ARGS__);                                      \
    if (awslc_fips_service_indicator_get_counter() > counter) { \
        approved = 1                                             \
    }                                             \
 while(0)
```
So that the caller doesn't have to know the service_id to check for an approved call.

Reply:
Changed to a macro definition and let all the service indicator functions to return `int`